Tiered Storage manual [HZ-4383] (#1047)
hazelcast/hz-docs: docs/modules/tiered-storage/pages/compactor.adoc

> = Compactor

Reviewer: There is a lot of info here that I'm not sure we need to detail. Instead, here's a suggested shortened section for the compactor:

The Tiered Storage compactor operates on each data partition to reclaim disk space by removing unused data.

When the configured threshold of available disk space is surpassed, compaction is initiated synchronously before and after a Map operation. To optimize the compaction process, the compactor scans the area with the highest garbage ratio to reclaim disk space efficiently. During this process, live entries discovered by the compactor are moved to memory, which may temporarily reduce the efficiency of the hot cache. 

WDYT @blazember, @Serdaro, @oliverhowell, @rebekah-lawrence - do we need much more?

We should keep the section on the fine-tuning.

Reviewer 2: Less is more. Seems good to me.

Reviewer 3: Addressed in https://github.com/hazelcast/hz-docs/pull/1047/commits/bb91f6376439c66396bc062a1d906f1a15b64397

The compactor scans the disk linearly, and moving the alive entries to the memory linearly too.
This means the IMap entries adjacent on the disk before compaction will be adjacent to each other after the compaction as well.
This can help with access patterns taking the entries in the same order.

=== Configuration Parameters

The compactor can be fine-tuned by the following properties:

|===
|Property |Default |Description

|`hazelcast.tiered.store.log.file.max.size.in.mb`
|8MB
|Maximum file size of a Tiered Storage file. It defines the working unit of the compactor.

|`hazelcast.tiered.store.partition.compactor.gc.threshold.per.map.partition.in.mb`
|The value set for `hazelcast.tiered.store.log.file.max.size.in.mb`
|The compactor on the partition threads is triggered only if there is at least this amount of garbage present on the partition.

|`hazelcast.tiered.store.compaction.starting.device.usage.percentage`
|0.9
|Triggers the compactor to run when the configured percentage of the configured device capacity is reached.

|`hazelcast.tiered.store.partition.compactor.scan.timeout.millis`
|10ms

> |Soft-limit for the compactor to run.

Reviewer 4: ```suggestion
|Max time a compaction can take in a single round
```

Author: @ahmetmircik It can be more than that, so maximum is misleading. WDYT about 

> Maximum time goal for a single compaction round. 

?

The alive entries the compactor finds get moved to the memory, temporarily decreasing the efficiency of the hot cache.

The compactor scans the disk linearly, and moving the alive entries to the memory linearly too.
This means the IMap entries adjacent on the disk before compaction will be adjacent to each other after the compaction as well.
This can help with access patterns taking the entries in the same order.

=== Configuration Parameters

The compactor can be fine-tuned by the following properties:

|===
|Property |Default |Description

|`hazelcast.tiered.store.log.file.max.size.in.mb`
|8MB
|Maximum file size of a Tiered Storage file. It defines the working unit of the compactor.

|`hazelcast.tiered.store.partition.compactor.gc.threshold.per.map.partition.in.mb`
|The value set for `hazelcast.tiered.store.log.file.max.size.in.mb`
|The compactor on the partition threads is triggered only if there is at least this amount of garbage present on the partition.

|`hazelcast.tiered.store.compaction.starting.device.usage.percentage`
|0.9
|Triggers the compactor to run when the configured percentage of the configured device capacity is reached.

> |`hazelcast.tiered.store.compactor.thread.count`

Reviewer 4: this is only relevant for concurrent compactors, maybe this line can be deleted.

= Compactor

== Overview

> The compactor used by Tiered Storage is a partitioned, synchronous, incremental, garbage-first, moving compactor, working on the disk only.

Author: Addressed in 06a2a152059bfb69e1201dfdbf7842c9b65f589f

> It's partitioned in the sense that compactions happen per Hazelcast partitions, offloaded from Hazelcast's partition operation threads to let operations assigned on the given partition operation thread to execute, including operations of the same partition, of different IMaps or data structures.

Author: Addressed in 06a2a152059bfb69e1201dfdbf7842c9b65f589f

> Tiered Storage compactors get triggered before and after the operations allocating data enough to reach the configured thresholds either per partition or globally.

Author: Addressed in 06a2a152059bfb69e1201dfdbf7842c9b65f589f

This means the compactor runs synchronously to the operations of the same partition and map.

Being incremental means that the compactor scans only a fraction of the used disk space at once to free the partition of the IMap as soon as possible.

The compactor of Tiered Storage is based on statistics.
When it comes to reclaim some space on the disk, it scans the area with the highest garbage ratio to be as efficient as possible, since working with garbage needs no further action to take.

> The alive entries the compactor finds get moved to the memory, temporarily decreasing the efficiency of the hot cache.

Author: Hot cache is basically the memory part of our tiered storage. We copy there all accessed map entries, so effectively Tiered Storage's memory part is a natural LRU cache.

The compactor scans the disk linearly, and moving the alive entries to the memory linearly too.
This means the IMap entries adjacent on the disk before compaction will be adjacent to each other after the compaction as well.
This can help with access patterns taking the entries in the same order.

=== Configuration Parameters

The compactor can be fine-tuned by the following properties:

|===
|Property |Default |Description

|`hazelcast.tiered.store.log.file.max.size.in.mb`
|8MB
|Maximum file size of a Tiered Storage file. It defines the working unit of the compactor.

|`hazelcast.tiered.store.partition.compactor.gc.threshold.per.map.partition.in.mb`
|The value set for `hazelcast.tiered.store.log.file.max.size.in.mb`

> |Makes the compactor on the partition threads getting triggered only if at least this amount of garbage is there on the partition.

Reviewer 3: ```suggestion
|The compactor on the partition threads is triggered only if there is at least this amount of garbage present on the partition.
```

Author: Addressed in 06a2a152059bfb69e1201dfdbf7842c9b65f589f

|`hazelcast.tiered.store.partition.compactor.gc.threshold.per.map.partition.in.mb`
|8MB

> |Makes the compactor on the partition threads getting triggered only if at least this amount of garbage is there on the partition.

Reviewer 3: This property is same as the previous one, with the default value change.

Author: Copy-paste error. Removed in 06a2a152059bfb69e1201dfdbf7842c9b65f589f

|`hazelcast.tiered.store.compaction.starting.device.usage.percentage`
|0.9

> |Forces the compactor running if the configured percentage of the configured device capacity is reached.

Reviewer 3: ```suggestion
|Triggers the compactor to run when the configured percentage of the configured device capacity is reached.
```

Author: Addressed in 06a2a152059bfb69e1201dfdbf7842c9b65f589f

|`hazelcast.tiered.store.compactor.thread.count`
|1
|The number of the compactor threads.

|`hazelcast.tiered.store.partition.compactor.scan.timeout.millis`
|10ms

> |Soft-limit for the compactor to runs.

Reviewer 3: ```suggestion
|Soft-limit for the compactor to run.
```

Author: Addressed in 06a2a152059bfb69e1201dfdbf7842c9b65f589f
